File: web/lib/custom-sample.ts
// Custom game samples for Advanced Stats: recompute ONE team's numbers from a
// chosen subset of its games, independently per team.
//
// The math and the data contract are documented in scripts/custom_sample.py.
// In short: rate stats are sums of raw per-game counts; opponent-adjusted edges
// are ridge-fit averages whose per-game contributions are published with the
// (fixed, official) opponent ratings baked in, so a subset of games gives that
// team's edge given those games; an all-games selection is the official value.
//
// Pure functions only -- no React, no fetching -- so they can be unit tested.

export type SampleGame = {
  g: string; // gameId
  w: number; // site-week
  o: string | null; // opponent name
  os: string | null; // opponent slug (null for non-FBS)
  oi: number | null; // opponent team id (null for non-FBS)
  fbs: boolean; // FBS-vs-FBS: contributes to the adjusted fit
  ha: "H" | "A" | "N";
  pf: number | null;
  pa: number | null;
  win: boolean;
  r: number[]; // raw counts, aligned with meta.rawFields
  s: number | null; // opponent's walk-forward SRS (SOS)
  a: Array<number | null>; // 6 numbers per spec: [numO, denO, uO, numD, denD, uD]
  x?: number[]; // optional turnover/penalty counts, aligned with meta.exFields
};

export type SampleMeta = {
  version: string;
  season: number;
  weekThrough: number;
  lambda: number;
  rampGames: number;
  rawFields: string[];
  specs: string[];
  exFields: string[];
  consts: Record<string, { mu: number; kOff: number; kDef: number }>;
};

export type SampleTeam = { slug: string; team: string; teamId: number; games: SampleGame[] };
export type TeamSampleData = { meta: SampleMeta; team: SampleTeam };

export type SampleResult = {
  included: number;
  total: number;
  wins: number;
  losses: number;
  /** Raw per-game counts summed over the chosen games, keyed like AdvancedRow.wk. */
  raw: Record<string, number>;
  /** Recomputed opponent-adjusted snapshot values, keyed like AdvancedRow. */
  adjusted: Record<string, number | null>;
  /** Turnover/penalty counts (Exploratory), or null when not exported for this team. */
  ex: Record<string, number> | null;
};

// Directly published adjusted edges: [AdvancedRow key, spec, fit side, decimals].
// Havoc's fitted sides are swapped on purpose (see build_real_data.py).
const DIRECT: Array<[string, string, "offense" | "defense", number]> = [
  ["offExp", "Explosive", "offense", 4],
  ["defExp", "Explosive", "defense", 4],
  ["offFin", "Finishing", "offense", 2],
  ["defFin", "Finishing", "defense", 2],
  ["offHavoc", "Havoc", "defense", 4],
  ["defHavoc", "Havoc", "offense", 4],
];

// Confidence-blended EPA / Success edges: [AdvancedRow key prefix, spec].
const BLENDED: Array<[string, string]> = [
  ["success", "Success"], ["epa", "EPA"], ["passEpa", "PassEPA"], ["rushEpa", "RushEPA"],
  ["passEpaDown1", "PassEPADown1"], ["passEpaDown2", "PassEPADown2"], ["passEpaDown3", "PassEPADown3"],
  ["rushEpaDown1", "RushEPADown1"], ["rushEpaDown2", "RushEPADown2"], ["rushEpaDown3", "RushEPADown3"],
  ["passSuccess", "PassSuccess"], ["rushSuccess", "RushSuccess"],
  ["passSuccessDown1", "PassSuccessDown1"], ["passSuccessDown2", "PassSuccessDown2"], ["passSuccessDown3", "PassSuccessDown3"],
  ["rushSuccessDown1", "RushSuccessDown1"], ["rushSuccessDown2", "RushSuccessDown2"], ["rushSuccessDown3", "RushSuccessDown3"],
];

/** AdvancedRow keys this module can recompute for a custom sample. */
export const CUSTOM_ADJUSTED_KEYS: string[] = [
  ...DIRECT.map(([key]) => key),
  ...BLENDED.flatMap(([prefix]) => [`${prefix}Adj`, `${prefix}AdjAllowed`]),
];

const round = (value: number, digits: number) => {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
};

function adjustmentConfidence(gamesPlayed: number, rampGames: number): number {
  if (gamesPlayed <= 1) return 0;
  if (rampGames <= 1) return 1;
  return Math.min(1, (gamesPlayed - 1) / (rampGames - 1));
}

/**
 * Recompute a team's numbers from the games in `included` (null = every game).
 * Opponent ratings stay at their official values; only THIS team's sample changes.
 */
export function computeSample(data: TeamSampleData, included: ReadonlySet<string> | null): SampleResult {
  const { meta, team } = data;
  const games = included ? team.games.filter((game) => included.has(game.g)) : team.games;
  const raw: Record<string, number> = {};
  meta.rawFields.forEach((field) => { raw[field] = 0; });
  let wins = 0;
  let losses = 0;
  let sosSum = 0;
  let sosCount = 0;

  type Acc = { aO: number; wO: number; obsO: number; aD: number; wD: number; obsD: number; nO: number; dO: number; nD: number; dD: number };
  const accs: Acc[] = meta.specs.map(() => ({ aO: 0, wO: 0, obsO: 0, aD: 0, wD: 0, obsD: 0, nO: 0, dO: 0, nD: 0, dD: 0 }));
  const exCounts: Record<string, number> = {};
  let exGames = 0;

  for (const game of games) {
    if (game.win) wins += 1; else losses += 1;
    meta.rawFields.forEach((field, index) => { raw[field] += game.r[index] ?? 0; });
    if (game.s !== null && game.s !== undefined) { sosSum += game.s; sosCount += 1; }
    meta.specs.forEach((_, si) => {
      const base = si * 6;
      const numO = game.a[base] ?? 0;
      const denO = game.a[base + 1] ?? 0;
      const uO = game.a[base + 2];
      const numD = game.a[base + 3] ?? 0;
      const denD = game.a[base + 4] ?? 0;
      const uD = game.a[base + 5];
      const acc = accs[si];
      acc.nO += numO; acc.dO += denO; acc.nD += numD; acc.dD += denD;
      if (game.fbs && uO !== null && uO !== undefined) { acc.aO += numO + denO * uO; acc.wO += denO; acc.obsO += 1; }
      if (game.fbs && uD !== null && uD !== undefined) { acc.aD += denD * uD - numD; acc.wD += denD; acc.obsD += 1; }
    });
    if (game.x) {
      exGames += 1;
      meta.exFields.forEach((field, index) => { exCounts[field] = (exCounts[field] ?? 0) + (game.x![index] ?? 0); });
    }
  }
  raw.opponentSrsSum = sosSum;
  raw.opponentSrsCount = sosCount;
  raw.wins = wins;
  raw.losses = losses;

  const specIndex = new Map(meta.specs.map((spec, index) => [spec, index]));
  const edge = (spec: string, side: "offense" | "defense"): number | null => {
    const si = specIndex.get(spec);
    const c = meta.consts[spec];
    if (si === undefined || !c) return null;
    const acc = accs[si];
    if (side === "offense") return acc.obsO ? (acc.aO - c.kOff) / (acc.wO + meta.lambda) : null;
    return acc.obsD ? (acc.aD - c.kDef) / (acc.wD + meta.lambda) : null;
  };

  const adjusted: Record<string, number | null> = {};
  for (const [key, spec, side, digits] of DIRECT) {
    if (!specIndex.has(spec)) continue;
    const value = edge(spec, side);
    adjusted[key] = value === null ? null : round(value, digits);
  }
  const confidence = adjustmentConfidence(games.length, meta.rampGames);
  for (const [prefix, spec] of BLENDED) {
    const si = specIndex.get(spec);
    const c = meta.consts[spec];
    if (si === undefined || !c) continue;
    const acc = accs[si];
    for (const [side, suffix] of [["offense", "Adj"], ["defense", "AdjAllowed"]] as const) {
      const e = edge(spec, side);
      const rawRate = side === "offense" ? (acc.dO > 0 ? acc.nO / acc.dO : null) : (acc.dD > 0 ? acc.nD / acc.dD : null);
      if (e === null || rawRate === null) { adjusted[prefix + suffix] = null; continue; }
      const unadjusted = side === "defense" ? c.mu - rawRate : rawRate - c.mu;
      adjusted[prefix + suffix] = round(confidence * e + (1 - confidence) * unadjusted, 4);
    }
  }

  return {
    included: games.length,
    total: team.games.length,
    wins,
    losses,
    raw,
    adjusted,
    ex: exGames > 0 && exGames === games.length ? exCounts : null,
  };
}

/**
 * Fail-closed guard: with every game selected, the recomputed adjusted values
 * must equal the official Advanced row (to the published rounding). If they do
 * not -- a stale artifact, a model change -- custom samples are disabled rather
 * than showing numbers that disagree with the official table.
 */
export function verifyParity(data: TeamSampleData, official: Record<string, unknown>): { ok: boolean; checked: number; maxDiff: number } {
  const all = computeSample(data, null);
  let checked = 0;
  let maxDiff = 0;
  let ok = true;
  for (const key of CUSTOM_ADJUSTED_KEYS) {
    const mine = all.adjusted[key];
    const theirs = official[key];
    if (mine === undefined) continue;
    if (mine === null && (theirs === null || theirs === undefined)) continue;
    if (typeof mine !== "number" || typeof theirs !== "number") { ok = false; continue; }
    const diff = Math.abs(mine - theirs);
    // Published values are rounded to 2 or 4 decimals; allow one last-digit flip.
    const tolerance = key === "offFin" || key === "defFin" ? 0.0101 : 0.000101;
    checked += 1;
    maxDiff = Math.max(maxDiff, diff);
    if (diff > tolerance) ok = false;
  }
  return { ok: ok && checked > 0, checked, maxDiff };
}

/** `michigan:4018.4019;iowa:4020` <-> { michigan: [...], iowa: [...] } */
export function serializeExclusions(excluded: Record<string, string[]>): string {
  return Object.entries(excluded)
    .filter(([, ids]) => ids.length > 0)
    .map(([slug, ids]) => `${slug}:${ids.join(".")}`)
    .join(";");
}

export function parseExclusions(value: string | null | undefined): Record<string, string[]> {
  const out: Record<string, string[]> = {};
  if (!value) return out;
  for (const part of value.split(";")) {
    const [slug, ids] = part.split(":");
    if (!slug || !/^[a-z0-9-]+$/.test(slug) || !ids) continue;
    const list = ids.split(".").filter((id) => /^\d{4,12}$/.test(id));
    if (list.length) out[slug] = [...new Set(list)];
  }
  return out;
}

File: scripts/custom_sample.py
"""Per-team, per-game export behind the Advanced page's "custom sample" feature.

A custom sample lets a user drop individual games from ONE team's Advanced
Stats (e.g. Michigan without Western Michigan, Iowa without Northern Iowa)
and recompute that team's numbers from the games that remain, independently
per team. This module publishes exactly the ingredients the browser needs so
the recomputation is exact and never touches Supabase per click.

How the published numbers are made -- and why they can be recomputed
---------------------------------------------------------------------
* "Rate" columns (YPP, success, explosive %, havoc %, pace, possession,
  field position, SOS, turnovers/penalties) are sums of raw per-game counts
  divided by each other. A subset of games is just a subset of the sums.
* Opponent-adjusted edges come from ``iterative_ratings.fit_metric_ratings``:
  a ridge fit of  y_g = mu + off(team) - def(opp)  over every FBS-vs-FBS
  team-game, observation weight w_g = the game's own denominator, ridge
  ``lambda`` = 50. At convergence a team's offense is

      off_i = sum_g w_g (y_g - mu + def_opp(g)) / (sum_g w_g + lambda)      (uncentered)

  and the published values are then centered. Writing a_g = w_g (y_g - mu_c +
  def_c[opp]) = num_g + den_g * u_g  with  u_g = def_c[opp] - mu_c  and
  centered (published) opponent ratings, every published offense edge is

      off_c,i = ( sum_g a_g - offenseMean * lambda ) / ( sum_g w_g + lambda )

  and symmetrically for defense (``a_g = den_g * u_g - num_g`` with
  ``u_g = mu_c + off_c[opp]``, ``defenseMean``). Restricting the sums to a
  chosen subset S of the team's games, while holding every opponent's
  published rating fixed, gives the team's edge *given those games*; with S =
  all games it is the published value. The three constants per spec (mu_c,
  offenseMean*lambda, defenseMean*lambda) are recorded next to the games.
* EPA / Success splits are then confidence-blended with the team's raw rate
  (``build_real_data.blend_edge``), using games played = |S|.

What is deliberately NOT recomputed: the headline ratings (Net/Off/Def, ASM)
and every ranking are official model outputs and stay official.
"""
from __future__ import annotations

from collections import defaultdict
from typing import Any

SAMPLE_VERSION = "advanced-custom-sample-v1"
LAMBDA = 50.0  # fit_metric_ratings default shrinkage used by fit_all_ratings()
# Ridge-fit specs whose per-team edges reach the Advanced page as snapshots.
ADJ_SPECS = (
    "Explosive", "Finishing", "Havoc",
    "Success", "EPA", "PassEPA", "RushEPA",
    "PassEPADown1", "PassEPADown2", "PassEPADown3",
    "RushEPADown1", "RushEPADown2", "RushEPADown3",
    "PassSuccess", "RushSuccess",
    "PassSuccessDown1", "PassSuccessDown2", "PassSuccessDown3",
    "RushSuccessDown1", "RushSuccessDown2", "RushSuccessDown3",
)
# Per-game turnover/penalty counts from the Exploratory export (2025+ data).
EX_FIELDS = (
    "games", "offensiveDrives", "opponentDrives", "passAttempts", "opponentPassAttempts",
    "turnovers", "interceptions", "lostFumbles", "turnoverEpaSum",
    "takeaways", "interceptionsForced", "fumbleRecoveries", "opponentTurnoverEpaSum",
    "offensivePenalties", "offensivePenaltyYards", "penaltiesDrawn", "penaltyYardsDrawn",
    "penaltiesForced", "penaltyYardsForced", "defensivePenalties", "defensivePenaltyYards",
)
# (published key prefix, spec, raw accumulator num/den field on the row) --
# mirrors build_real_data.EPA_SUCCESS_METRICS; a unit test pins the two together.
BLENDED = (
    ("success", "Success"), ("epa", "EPA"), ("passEpa", "PassEPA"), ("rushEpa", "RushEPA"),
    ("passEpaDown1", "PassEPADown1"), ("passEpaDown2", "PassEPADown2"), ("passEpaDown3", "PassEPADown3"),
    ("rushEpaDown1", "RushEPADown1"), ("rushEpaDown2", "RushEPADown2"), ("rushEpaDown3", "RushEPADown3"),
    ("passSuccess", "PassSuccess"), ("rushSuccess", "RushSuccess"),
    ("passSuccessDown1", "PassSuccessDown1"), ("passSuccessDown2", "PassSuccessDown2"), ("passSuccessDown3", "PassSuccessDown3"),
    ("rushSuccessDown1", "RushSuccessDown1"), ("rushSuccessDown2", "RushSuccessDown2"), ("rushSuccessDown3", "RushSuccessDown3"),
)
# Directly-published (non-blended) snapshot edges: (key, spec, fit side). Havoc's
# fitted sides are swapped on purpose -- see build_real_data.py.
DIRECT = (
    ("offExp", "Explosive", "offense", 4), ("defExp", "Explosive", "defense", 4),
    ("offFin", "Finishing", "offense", 2), ("defFin", "Finishing", "defense", 2),
    ("offHavoc", "Havoc", "defense", 4), ("defHavoc", "Havoc", "offense", 4),
)
EPA_ADJUSTMENT_RAMP_GAMES = 5


def _num(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool) and v == v


def _f(v: Any) -> float:
    return float(v) if _num(v) else 0.0


def _r(v: float, digits: int) -> float | int:
    out = round(float(v), digits)
    return int(out) if out == int(out) and abs(out) < 1e15 else out


def _spec_fields() -> dict[str, tuple[str, str]]:
    from cfb_analytics.analytics.iterative_ratings import SPECS

    return {name: (numerator, denominator) for name, numerator, denominator in SPECS}


def _identity(row: dict[str, Any]) -> tuple[str, str]:
    return str(row.get("gameId") or row.get("game_id")), str(row["team"])


def _load_exploratory(year: int) -> dict[tuple[str, str], list[float]]:
    try:
        from export_exploratory_data import load_exploratory_rows
    except Exception:  # noqa: BLE001 - optional research export
        return {}
    try:
        rows = load_exploratory_rows(year)
    except Exception:  # noqa: BLE001
        return {}
    return {
        (str(row["gameId"]), str(row["team"])): [_r(_f(row.get(field)), 6) for field in EX_FIELDS]
        for row in rows
    }


def build_artifact(*, year: int, week_through: int, games: list[dict[str, Any]], metric_ratings: dict[str, Any],
                   opp_srs_by_key: dict[tuple[str, str], float], poss_seconds: dict[str, Any]) -> dict[str, Any]:
    """The full-season, all-team artifact (one JSON per season, private)."""
    from build_game_logs import extract_raw_fields

    fields = _spec_fields()
    consts: dict[str, dict[str, float]] = {}
    for spec in ADJ_SPECS:
        fit = metric_ratings.get(spec) or {}
        if not _num(fit.get("leagueMean")) or not fit.get("offense"):
            continue
        shrink = float(fit.get("shrinkage", LAMBDA))
        if shrink != LAMBDA:
            raise RuntimeError(f"{spec}: unexpected shrinkage {shrink}; custom-sample math assumes {LAMBDA}")
        consts[spec] = {
            "mu": float(fit["leagueMean"]),
            "kOff": float(fit["offenseMean"]) * LAMBDA,
            "kDef": float(fit["defenseMean"]) * LAMBDA,
        }
    specs = [s for s in ADJ_SPECS if s in consts]

    by_game: dict[str, dict[str, dict[str, Any]]] = defaultdict(dict)
    meta: dict[str, dict[str, Any]] = {}
    for row in games:
        gid, team = _identity(row)
        by_game[gid][team] = row
        meta[team] = {"slug": row.get("team_slug"), "teamId": row.get("team_id")}

    exploratory = _load_exploratory(year)
    raw_fields: list[str] | None = None
    teams: dict[str, dict[str, Any]] = {}
    for row in sorted(games, key=lambda r: (r["team"], r["_siteWeek"], _identity(r)[0])):
        gid, team = _identity(row)
        opp = row.get("opponent")
        opp_row = by_game[gid].get(str(opp))
        is_fbs = row.get("opponent_classification") == "fbs" and opp_row is not None
        raw = extract_raw_fields(row, poss_seconds)
        if raw_fields is None:
            raw_fields = list(raw.keys())
        entry: dict[str, Any] = {
            "g": gid, "w": int(row["_siteWeek"]), "o": opp,
            "os": (meta.get(opp) or {}).get("slug"), "oi": (meta.get(opp) or {}).get("teamId"),
            "fbs": bool(is_fbs),
            "ha": "N" if row.get("neutral_site") else ("H" if row.get("home_away") == "home" else "A"),
            "pf": row.get("points_for"), "pa": row.get("points_against"),
            "win": bool(row.get("win")),
            "r": [_r(_f(raw[k]), 4) for k in raw_fields],
            "s": _r(opp_srs_by_key[(gid, team)], 4) if (gid, team) in opp_srs_by_key else None,
        }
        adj: list[Any] = []
        for spec in specs:
            nf, df = fields[spec]
            own_n, own_d = _f(row.get(nf)), _f(row.get(df))
            if spec == "Havoc":
                # Havoc's fitted defense side is the opponent's own havoc fields.
                opp_n, opp_d = (_f(opp_row.get(nf)), _f(opp_row.get(df))) if opp_row else (0.0, 0.0)
            else:
                # Same counts as the opponent's own numerator/denominator in an
                # FBS-vs-FBS game (asserted by the tests), but ALSO present for an
                # FCS opponent, whose row is dropped -- the raw blend needs them.
                opp_n, opp_d = _f(row.get(nf + "Allowed")), _f(row.get(df + "Allowed"))
            # Own offense vs opponent defense; own defense vs opponent offense.
            u_off = u_def = None
            if is_fbs:
                fit = metric_ratings[spec]
                d_opp, o_opp = fit["defense"].get(opp), fit["offense"].get(opp)
                mu = consts[spec]["mu"]
                if d_opp is not None and own_d > 0:
                    u_off = _r(d_opp - mu, 8)
                if o_opp is not None and opp_d > 0:
                    u_def = _r(mu + o_opp, 8)
            adj += [_r(own_n, 6), _r(own_d, 6), u_off, _r(opp_n, 6), _r(opp_d, 6), u_def]
        entry["a"] = adj
        ex = exploratory.get((gid, team))
        if ex is not None:
            entry["x"] = ex
        team_key = f"t{meta[team]['teamId']}"
        teams.setdefault(team_key, {"slug": meta[team]["slug"], "team": team, "teamId": meta[team]["teamId"], "games": []})["games"].append(entry)

    return {
        # `meta` and `teams` are separate top-level keys so the API route can read
        # one team with a PostgREST JSON-path select (payload->teams->t<id>) instead
        # of pulling the whole season out of Supabase.
        "meta": {
            "version": SAMPLE_VERSION, "season": year, "weekThrough": week_through, "lambda": LAMBDA,
            "rampGames": EPA_ADJUSTMENT_RAMP_GAMES,
            "rawFields": raw_fields or [], "specs": specs, "exFields": list(EX_FIELDS), "consts": consts,
        },
        "teams": teams,
    }


# ---------------------------------------------------------------------------
# Reference implementation (mirrors web/lib/custom-sample.ts). Used by the
# parity test against the values build_year() itself publishes.
# ---------------------------------------------------------------------------
def _adjust_confidence(games_played: int, ramp: int = EPA_ADJUSTMENT_RAMP_GAMES) -> float:
    if games_played <= 1:
        return 0.0
    return min(1.0, (games_played - 1) / (ramp - 1))


def reference_adjusted(artifact: dict[str, Any], team_key: str, excluded: set[str] | None = None) -> dict[str, float | None]:
    excluded = excluded or set()
    meta = artifact["meta"]
    specs = meta["specs"]
    chosen = [g for g in artifact["teams"][team_key]["games"] if g["g"] not in excluded]
    out: dict[str, float | None] = {}
    sums: dict[str, dict[str, float]] = {}
    for si, spec in enumerate(specs):
        acc = {"aO": 0.0, "wO": 0.0, "aD": 0.0, "wD": 0.0, "nO": 0.0, "dO": 0.0, "nD": 0.0, "dD": 0.0, "obsO": 0, "obsD": 0}
        for g in chosen:
            n_o, d_o, u_o, n_d, d_d, u_d = g["a"][si * 6:(si + 1) * 6]
            acc["nO"] += n_o; acc["dO"] += d_o; acc["nD"] += n_d; acc["dD"] += d_d
            if g["fbs"] and u_o is not None:
                acc["aO"] += n_o + d_o * u_o; acc["wO"] += d_o; acc["obsO"] += 1
            if g["fbs"] and u_d is not None:
                acc["aD"] += d_d * u_d - n_d; acc["wD"] += d_d; acc["obsD"] += 1
        sums[spec] = acc
    n_games = len(chosen)

    def edge(spec: str, side: str) -> float | None:
        c, acc = meta["consts"][spec], sums[spec]
        if side == "offense":
            return (acc["aO"] - c["kOff"]) / (acc["wO"] + meta["lambda"]) if acc["obsO"] else None
        return (acc["aD"] - c["kDef"]) / (acc["wD"] + meta["lambda"]) if acc["obsD"] else None

    for key, spec, side, digits in DIRECT:
        if spec in sums:
            v = edge(spec, side)
            out[key] = None if v is None else round(v, digits)
    for prefix, spec in BLENDED:
        if spec not in sums:
            continue
        c, acc = meta["consts"][spec], sums[spec]
        for side, suffix in (("offense", "Adj"), ("defense", "AdjAllowed")):
            e = edge(spec, side)
            raw = (acc["nO"] / acc["dO"] if acc["dO"] else None) if side == "offense" else (acc["nD"] / acc["dD"] if acc["dD"] else None)
            if e is None or raw is None:
                out[prefix + suffix] = None
                continue
            conf = _adjust_confidence(n_games)
            unadjusted = (c["mu"] - raw) if side == "defense" else (raw - c["mu"])
            out[prefix + suffix] = round(conf * e + (1.0 - conf) * unadjusted, 4)
    return out
